src/provider/model/Model.ts

```diff
@@ -128,21 +131,32 @@ export class Model {
     }
 
     private async onWindowCreated(identity: Identity, manifestUrl: string): Promise<void> {
+        // Registration is asynchronous and sensitive to race conditions. We use a deferred promise
+        // to signal to other sensitive operations that it is safe to proceed.
+        const [pendingRegistration, clearPending] = deferredPromise();
+        this._pendingRegistrations.set(getId(identity), pendingRegistration);
+
         const apps = await this._directory.getAllApps();
         const appInfoFromDirectory = apps.find(app => app.manifest.startsWith(manifestUrl));
 
+        const id: string = getId(identity);
         if (!appInfoFromDirectory) {
```

Comment:
Could this block be a single `if` statement - `if (appInfoFromDirectory && !this._windowsById[id])`?

Or in any case, I'd suggest have each path fall through to the clearPending and delete at the bottom, rather than `return`ing. This is one of my "I can't explain this properly, but the graph in my head looks neater if it's done that way" things

Reply:
I thought I had done that, but apparently it got messed up in git somewhere.

Reply:
I was also keen to keep the "Ignoring window" comment as it helped with debugging.